allennlp/data/iterators/cached_iterator.py

```diff
@@ -0,0 +1,45 @@
+import logging
+import random
+from typing import Iterable, Dict, List
+from overrides import overrides
+from allennlp.data.instance import Instance
+from allennlp.data.iterators.bucket_iterator import BucketIterator
+from allennlp.data.iterators.data_iterator import DataIterator
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@DataIterator.register("cached")
+class CachedIterator(BucketIterator):
+    """
+    An extension of the `BucketIterator` that caches its batches between epochs. Batches are built
+    only once during the first epoch, then the same batches are used again in consecutive epochs.
+    Using this iterator saves run time but costs more memory to store the cached batches. It should
+    be used only if the dataset can fit in memory. This iterator shouldn't be used with the lazy
+    mode.
+
+    Parameters:
+    ----------
+        See :class:`BucketIterator`.
+    """
+
+    @overrides
+    def _yield_one_epoch(self, instances: Iterable[Instance], shuffle: bool, cuda_device: int, for_training: bool):
+        if not hasattr(self, 'cached_batches'):
+            # instances id -> list[batches]
+            self.cached_batches: Dict[int, List] = {}  # pylint: disable=attribute-defined-outside-init
```

Comment:
and then I would just define this in the constructor

Reply:
do you know how to define a constructor for a subclass that takes the same arguments of the parent class without explicitly listing them ? I don't want to copy-paste them from `BucketIterator.__init__`

Reply:
```
class SubClass( Super ):
   def __init__(self, *args, **kwargs):
       super(self).__init__(*args, **kwargs)

```